os/fs/smartfs/smartfs_smart.c

```diff
@@ -712,24 +716,26 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
 				goto errout_with_semaphore;
 			}
 
-			/* If file is modified and more data remains to be appended to file, but no next sector is available,
-			 * do not update sf->currsector and curroffset. Will be handled when buffer data is synced.
+			/* If file is modified and more data remains to be appended to file,
+			 * but no next sector is available, do not update sf->currsector and curroffset.
+			 * This will be handled when buffer data is synced.
 			 */
 			if (SMARTFS_NEXTSECTOR(header) != 0xFFFF) {
-				sf->currsector = SMARTFS_NEXTSECTOR(header);
-				
 				/* Now get the chained sector info and reset the offset */
+				sf->currsector = SMARTFS_NEXTSECTOR(header);
 				sf->curroffset = size;
+
 #ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-				/* Read the next sector if more bytes remain in the write buffer */
-				if (buflen > 0) {
-					smartfs_setbuffer(&readwrite, sf->currsector, 0,\
-						 fs->fs_llformat.availbytes, (uint8_t *)sf->buffer);
-					ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
-					if (ret < 0) {
-						fdbg("Error %d reading sector %d header\n", ret, sf->currsector);
-						goto errout_with_semaphore;
-					}
+				/* Read the next sector if we have reached the end of the current sector.
+				 * Future file ops need this data to be present since buffer updation does not happen
+				 * automatically/otherwise
+				 */
+				smartfs_setbuffer(&readwrite, sf->currsector, 0,\
+					 fs->fs_llformat.availbytes, (uint8_t *)sf->buffer);
+				ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
```

Comment:
will it not be an overhead in case no next sector available case?

Reply:
Updated with an alternate solution to avoid overhead.
Please review.